apps/api_web/lib/api_web/event_stream/supervisor.ex

```diff
@@ -2,52 +2,68 @@ defmodule ApiWeb.EventStream.Supervisor do
   @moduledoc """
   Supervisor for the infrastructure needed for event streaming.
 
-  - ServerRegistry - mapping controller/params pairs to an EventStream.Server
+  - ServerRegistry - mapping controller/params pairs to a DiffServer
+  - SubscriberRegistry - bag of all processes that are subscribed to a DiffServer
   - ServerSupervisor - DynamicSupervisor for managing the children
   """
+
+  alias ApiWeb.EventStream.{DiffServer, ServerRegistry, ServerSupervisor}
+
   def start_link do
     Supervisor.start_link(
       [
-        {Registry, name: ApiWeb.EventStream.ServerRegistry, keys: :unique},
-        {DynamicSupervisor, name: ApiWeb.EventStream.ServerSupervisor, strategy: :one_for_one}
+        {Registry, name: ServerRegistry, keys: :unique},
+        {Registry, name: SubscriberRegistry, keys: :duplicate},
+        {DynamicSupervisor, name: ServerSupervisor, strategy: :one_for_one}
       ],
       strategy: :one_for_all
     )
   end
 
   @doc """
-  Returns a {:ok, pid} tuple for an EventStream.Server for the given connection/module.
-
-  If one already exists, the same one will be returned.
+  Subscribes the current process to the appropriate DiffServer for the given connection/module,
+  starting it if it doesn't exist.
   """
-  def server_child(conn, module) do
+  @spec server_subscribe(Plug.Conn.t(), module) :: DynamicSupervisor.on_start_child()
+  def server_subscribe(conn, module) do
     case start_child(conn, module) do
-      {:ok, pid} ->
-        :ok = ApiWeb.EventStream.DiffServer.subscribe(pid)
-        {:ok, pid}
+      {:ok, pid} -> do_subscribe(pid)
+      {:error, {:already_started, pid}} -> do_subscribe(pid)
+      {:error, _} = error -> error
+    end
+  end
 
-      {:error, {:already_started, pid}} ->
-        :ok = ApiWeb.EventStream.DiffServer.subscribe(pid)
-        {:ok, pid}
+  @doc "Unsubscribes the current process from the given DiffServer."
+  @spec server_unsubscribe(pid) :: :ok
+  def server_unsubscribe(pid) do
+    :ok = Registry.unregister(SubscriberRegistry, :subscribers)
+    :ok = DiffServer.unsubscribe(pid)
+  end
 
-      {:error, _} = error ->
-        error
-    end
+  @doc "Sends the given message to all processes subscribed to a DiffServer."
+  @spec subscriber_dispatch(any) :: :ok
+  def subscriber_dispatch(message) do
```

Comment:
Do you need to talk to the clients of the `DiffServer`s? If you shut the DiffServer's normally, their clients should receive a `:DOWN` message that they can handle by gracefully exiting.

Reply:
It currently works the other way around: DiffServers monitor their clients (so it can remove the client from its list and terminate if it has no clients anymore).

I don't think there would be anything stopping the clients from also monitoring their DiffServer (?), in which case we wouldn't need the registry of clients... but then we're also leaning a bit more on implementation details. If the goal is to close the HTTP connections, why do that by terminating a different process you know the connection processes are monitoring (a mostly irrelevant detail), when you could notify them directly?

We probably should have connection processes monitor their DiffServer, though; if it goes down unexpectedly, for some reason other than a shutdown, we don't want those connections to stay open but stop delivering new data.

Reply:
...as I think about this more it really feels like a "conversation" that should be happening exclusively between the Canary and the HTTP processes. The one exists to notify the others of a shutdown; there's no reason for even the DiffServer Supervisor to be involved. I'll try a more isolated approach.

I did try using the "terminate the diffservers" approach for comparison: it's in `cfg-close-event-streams-by-terminating`

Reply:
Eh, it turns out this was harder than it looked. This has far exceeded its timebox already so I'm going to leave it where it is for now — unless you're strongly of the opinion that the `by-terminating` approach is better and we should use that, in which case I will swap it in.

Reply:
What do you think is better about this approach, compared to the `by-terminating` approach? To me, the `by-terminating` approach looks much cleaner.

Reply:
This sums it up I think:

> [the by-terminating approach is] leaning a bit more on implementation details. If the goal is to close the HTTP connections, why do that by terminating a different process you know the connection processes are monitoring (a mostly irrelevant detail), when you could notify them directly?

I would agree it's an approach that requires fewer changes and less code overall, but it's also more tied to an implementation detail. So I think it's a trade-off, not unambiguously better.

Reply:
You need to communicate with the `DiffServer` supervisor either way: at that point the implementation of shutting down the connections is out of the hands of `Canary` and we can go with whatever is easiest (the other approach). 

If we do need to go back to "communicate with each HTTP connection", we can always update `terminate_servers` to send the shutdown messages at that point.

Reply:
> You need to communicate with the `DiffServer` supervisor either way

This is sort of why I thought it would be better to not do that at all, by having the Canary maintain the registry and having the connection processes "subscribe" directly to it for shutdown notifications. But agree that the approach already written in `by-terminating` is fine, in the interests of not churning further on this 😄